File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateK(self, head, k):
        tail = head
        prev = None
        while(head != None and k>0):
            temp = head.next
            head.next = prev
            prev = head
            head = temp
            k -= 1
        newHead = prev
        return [newHead, tail, head]
    
    def lengthOfList(self, head):
        length = 0
        
        while(head != None):
            head = head.next
            length += 1
        return length
    
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head
        oldHead = head
        k = right-left+1
        tempLeft = left
        length = self.lengthOfList(head)
        
        while(head != None and left>1):
            tempHead = head
            head = head.next
            left -= 1
        newHead, tail, oldTail = self.rotateK(head, k)
        
        tail.next = oldTail
        if k == length or tempLeft == 1:
            return newHead
        else:
            tempHead.next = newHead
        return oldHead
        

File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(left==right || head == null || head.next == null) return head;
        int tempLeft = left-1;
        ListNode oldMainHead = head;
        ListNode oldHead = head;
        while(head != null && tempLeft-- > 0) {
            oldHead = head;
            head = head.next;
        }
        int k = right-left+1;
        ListNode[] rotateVars = rotateKTimes(head, k);
        ListNode newTail = rotateVars[0];
        ListNode newHead = rotateVars[1];
        ListNode oldTail = rotateVars[2];
        oldHead.next = newHead;
        newTail.next = oldTail;
        if(left == 1) {
            return newHead;
        } else {
            return oldMainHead;
        }
    }
    public ListNode[] rotateKTimes(ListNode node, int k) {
        ListNode tail = node;
        ListNode prev = null;
        while(k-- > 0 && node != null) {
            ListNode next = node.next;
            node.next = prev;
            prev = node;
            node = next;
        }
        return new ListNode[]{tail, prev, node};
    }
}